Source/Dafny/Compiler.cs

```diff
@@ -198,19 +211,33 @@ protected interface IClassWriter {
     /// </summary>
     protected abstract TargetWriter DeclareLocalVar(string name, Type/*?*/ type, Bpl.IToken/*?*/ tok, TargetWriter wr);
     protected virtual void DeclareOutCollector(string collectorVarName, TargetWriter wr) { }  // called only for return-style calls
+    protected virtual void DeclareSpecificOutCollector(string collectorVarName, TargetWriter wr, int outCount, List<Type> types, Method m) {DeclareOutCollector(collectorVarName, wr); } // for languages that don't allow "let" or "var" expressions
     protected virtual bool UseReturnStyleOuts(Method m, int nonGhostOutCount) => false;
+    protected virtual BlockTargetWriter EmitMethodReturns(Method m, BlockTargetWriter wr) { return wr; } // for languages that need explicit return statements not provided by Dafny
     protected virtual bool SupportsMultipleReturns { get => false; }
     protected virtual bool NeedsCastFromTypeParameter { get => false; }
+    protected virtual bool SupportsAmbiguousTypeDecl { get => true; }
+    protected virtual bool FieldsInTraits { get => true; } // True if language's "trait" equivalent allows for non-final field declarations, false otherwise
+    protected bool MemberSelectObjIsTrait; // Example: if MemberSelect expression is T.x, this will be true if T is a trait, false otherwise
```

Comment:
Which `MemberSelect` expression is this referring to?

This would be better as some kind of parameter to `EmitMemberSelect` and `EmitAssignment`.